File: examples/qa-bot/bot.ts
/**
 * QA Bot — Example application using the Ash SDK.
 *
 * Local usage:
 *   const bot = new QABot('http://localhost:4100');
 *   await bot.setup('qa-bot', './agent');
 *   const answer = await bot.ask('What is 2+2?');
 *   await bot.teardown();
 *
 * Remote usage (EC2/GCP):
 *   const bot = new QABot('http://<ec2-ip>:4100', 'your-api-key');
 *   await bot.resume('qa-bot', 'session-id-here');
 *   const answer = await bot.ask('What did we talk about?');
 */

import { AshClient, extractTextFromEvent } from '@ash-ai/sdk';
import type { Session } from '@ash-ai/sdk';

export interface BotMessage {
  role: 'user' | 'assistant';
  content: string;
}

export class QABot {
  private client: AshClient;
  private agentName: string | null = null;
  private session: Session | null = null;
  private history: BotMessage[] = [];

  constructor(serverUrl: string, apiKey?: string) {
    this.client = new AshClient({ serverUrl, apiKey });
  }

  /** Deploy agent and create a new session. */
  async setup(agentName: string, agentPath: string): Promise<Session> {
    await this.client.deployAgent(agentName, agentPath);
    this.agentName = agentName;
    this.session = await this.client.createSession(agentName);
    this.history = [];
    return this.session;
  }

  /** Resume an existing session by ID. */
  async resume(agentName: string, sessionId: string): Promise<Session> {
    this.agentName = agentName;
    this.session = await this.client.resumeSession(sessionId);
    this.history = [];
    return this.session;
  }

  /** List resumable sessions for an agent. */
  async listSessions(agentName?: string): Promise<Session[]> {
    const sessions = await this.client.listSessions(agentName ?? this.agentName ?? undefined);
    return sessions.filter(
      (s) => s.status === 'active' || s.status === 'paused' || s.status === 'error',
    );
  }

  /** Send a question and collect the streamed response. */
  async ask(question: string): Promise<string> {
    if (!this.session) throw new Error('Call setup() or resume() first');

    this.history.push({ role: 'user', content: question });

    const parts: string[] = [];
    for await (const event of this.client.sendMessageStream(this.session.id, question)) {
      if (event.type === 'message') {
        const text = extractTextFromEvent(event.data);
        if (text) parts.push(text);
      }
    }

    const content = parts.join('');

    this.history.push({ role: 'assistant', content });
    return content;
  }

  /** Get conversation history (local to this bot instance). */
  getHistory(): BotMessage[] {
    return [...this.history];
  }

  /** Get current session info from the server. */
  async getSessionInfo(): Promise<Session | null> {
    if (!this.session) return null;
    return this.client.getSession(this.session.id);
  }

  /** Get current session ID. */
  getSessionId(): string | null {
    return this.session?.id ?? null;
  }

  /** End the session and clean up. */
  async teardown(): Promise<void> {
    if (this.session) {
      await this.client.endSession(this.session.id);
      this.session = null;
    }
    if (this.agentName) {
      await this.client.deleteAgent(this.agentName);
      this.agentName = null;
    }
    this.history = [];
  }

  /** Check if Ash server is healthy. */
  async isHealthy(): Promise<boolean> {
    try {
      const h = await this.client.health();
      return h.status === 'ok';
    } catch {
      return false;
    }
  }
}

File: examples/python-bot/bot.py
#!/usr/bin/env python3
"""
Example: Python bot using the Ash Python SDK.

Deploys an agent, creates a session with SDK parity options
(system_prompt, model, effort, max_turns), has a multi-turn
conversation with SSE streaming, then cleans up.

Usage:
    python bot.py
    ASH_SERVER_URL=http://remote:4100 python bot.py
"""

import os
import sys
from pathlib import Path

# Allow running from the repo without installing the SDK package
sdk_path = Path(__file__).resolve().parent.parent.parent / "packages" / "sdk-python"
sys.path.insert(0, str(sdk_path))

from ash_sdk import AshClient
from ash_sdk.streaming import MessageEvent, TextDeltaEvent, DoneEvent, ErrorEvent


def extract_text(data: dict) -> str:
    """Extract text content from an SDK assistant message."""
    if data.get("type") != "assistant":
        return ""
    content = data.get("message", {}).get("content", [])
    if isinstance(content, str):
        return content
    if isinstance(content, list):
        return "".join(
            block.get("text", "") for block in content if block.get("type") == "text"
        )
    return ""


def main():
    server_url = os.environ.get("ASH_SERVER_URL", "http://localhost:4100")
    agent_dir = str(Path(__file__).resolve().parent / "agent")
    agent_name = "python-bot"

    client = AshClient(server_url)

    # Check server health
    print(f"Connecting to {server_url}...")
    try:
        health = client.health()
        print(f"Server status: {health['status']} (uptime: {health['uptime']}s)")
    except Exception as e:
        print(f"Error: Cannot reach server at {server_url}: {e}")
        sys.exit(1)

    # Deploy agent
    print(f"\nDeploying agent '{agent_name}' from {agent_dir}...")
    agent = client.deploy_agent(agent_name, agent_dir)
    print(f"Deployed: {agent.name} v{agent.version}")

    # Create session with SDK parity options
    print("\nCreating session with custom settings...")
    session = client.create_session(
        agent_name,
        system_prompt="You are a concise programming tutor. Answer in 2-3 sentences max.",
        permission_mode="bypassPermissions",
        allowed_tools=["Read", "Write", "Bash"],
    )
    print(f"Session: {session.id} (status: {session.status})")

    # Multi-turn conversation with per-message options
    questions = [
        "What is a closure in programming?",
        "Give me a one-line example in Python.",
        "Thanks! How about in JavaScript?",
    ]

    try:
        for i, question in enumerate(questions, 1):
            print(f"\n--- Turn {i} ---")
            print(f"You: {question}")
            print("Bot: ", end="", flush=True)

            for event in client.send_message_stream(
                session.id,
                question,
                max_turns=1,
            ):
                if isinstance(event, MessageEvent):
                    text = extract_text(event.data)
                    if text:
                        print(text, end="", flush=True)
                elif isinstance(event, TextDeltaEvent):
                    # Incremental text when include_partial_messages=True
                    pass
                elif isinstance(event, ErrorEvent):
                    print(f"\n[Error: {event.error}]")
                elif isinstance(event, DoneEvent):
                    print()  # newline after streamed response

    finally:
        # Clean up
        print("\nEnding session...")
        ended = client.end_session(session.id)
        print(f"Session ended (status: {ended.status})")

        print("Cleaning up agent...")
        client.delete_agent(agent_name)
        print("Done.")


if __name__ == "__main__":
    main()
